src/arch/x86/kernel/boot_sys.c

```diff
@@ -628,6 +630,18 @@ static BOOT_CODE bool_t try_boot_sys_mbi2(
         if (tag->type == MULTIBOOT2_TAG_CMDLINE) {
             char const *const cmdline = (char const * const)(behind_tag);
             cmdline_parse(cmdline, &cmdline_opt);
+        } else if (tag->type == MULTIBOOT2_TAG_EFI32) {
+            if (tag->size - sizeof(*tag) == sizeof(uint32_t)) {
+                boot_state.efi_ptr32 = *(uint32_t *)behind_tag;
+                printf("32-bit EFI system table at physical address=0x%x\n",
+                       boot_state.efi_ptr32);
+            }
```

Comment:
If you only keep one 64-bit value and if bootloaders are allowed to pass both, you probably need an extra check to see if the 64-bit value wasn't set yet before accidentally overwriting it with the 32-bit value here (assuming that's the behaviour you want).

Reply:
I'm not sure if the bootloader is allow to pass both, but my concern is sometimes 64-bit platforms can ship with 32-bit firmware (and 64-bit firmware can boot 32-bit OS, at least on x86), so I think the safest option is to pass the information from the bootloader verbatim, and let the user space task decide what to do with it. I've modified the commit to pass both if the bootloader has specified both (and fixed a bug where the extra bootinfo size is calculated incorrectly).

Reply:
That just confirms that the bootloader only passes one value, in which case there is no point is passing both. Just always pass the value as 64-bits to user space, like I suggested above. User space can check whether the top 32-bits are zero to see if it's a 32-bit value or a 64-bit one.

Reply:
What about having one SEL4_BOOTINFO_HEADER_EFI_SYSTBL_PTR tag and always encode it 64-bits, but separately convey the information that it's a 64-bit machine running 32-bit OS?

Reply:
Of course it's functionally the same. But instead of having to check a random size of a random field to determine what mode GRUB booted in, the info is told explicitly to the user instead. However, I suppose the only place people need to know is when accessing EFI tables, so having it here makes sense.